chain/network/src/peer_manager/tests/sync_snapshot_hosts.rs

```diff
@@ -0,0 +1,282 @@
+    let clock = clock.clock();
+    let clock = &clock;
+
+    let pm = peer_manager::testonly::start(
+        clock.clone(),
+        near_store::db::TestDB::new(),
+        chain.make_config(rng),
+        chain.clone(),
+    )
+    .await;
+
+    tracing::info!(target:"test", "Connect a peer, expect initial sync to be empty.");
+    let peer1_config = chain.make_config(rng);
+    let mut peer1 =
+        pm.start_inbound(chain.clone(), peer1_config.clone()).await.handshake(clock).await;
+    let empty_sync_msg = peer1.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(empty_sync_msg.hosts, vec![]);
+
+    tracing::info!(target:"test", "Connect two more peers.");
+    let peer2_config = chain.make_config(rng);
+    let mut peer2 =
+        pm.start_inbound(chain.clone(), peer2_config.clone()).await.handshake(clock).await;
+    let empty_sync_msg = peer2.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(empty_sync_msg.hosts, vec![]);
+
+    let mut peer3 =
+        pm.start_inbound(chain.clone(), chain.make_config(rng)).await.handshake(clock).await;
+    let empty_sync_msg = peer3.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(empty_sync_msg.hosts, vec![]);
+
+    tracing::info!(target:"test", "Send a SyncSnapshotHosts message from peer1, make sure that all peers receive it.");
+
+    let info1 =
+        make_snapshot_host_info(&peer1_config.node_id(), 123, vec![0, 1], &peer1_config.node_key);
+
+    peer1
+        .send(PeerMessage::SyncSnapshotHosts(SyncSnapshotHosts { hosts: vec![info1.clone()] }))
+        .await;
+
+    let got1 = peer1.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(got1.hosts, vec![info1.clone()]);
+
+    let got2 = peer2.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(got2.hosts, vec![info1.clone()]);
+
+    let got3 = peer3.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(got3.hosts, vec![info1.clone()]);
+
+    tracing::info!(target:"test", "Connect another peer, make sure that it receives the correct information on joining.");
+    let mut peer4 =
+        pm.start_inbound(chain.clone(), chain.make_config(rng)).await.handshake(clock).await;
+    let peer4_sync_msg = peer4.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(peer4_sync_msg.hosts, vec![info1.clone()]);
+
+    tracing::info!(target:"test", "Publish another piece of snapshot information, check that it's also broadcasted.");
+    let info2 =
+        make_snapshot_host_info(&peer2_config.node_id(), 11212, vec![3], &peer2_config.node_key);
+
+    peer2
+        .send(PeerMessage::SyncSnapshotHosts(SyncSnapshotHosts { hosts: vec![info2.clone()] }))
+        .await;
+
+    let got1 = peer1.events.recv_until(take_sync_snapshot_msg).await;
+    assert_eq!(got1.hosts, vec![info2.clone()]);
+
+    tracing::info!(target:"test", "Connect another peer, check that it receieves all of the published information.");
+
```

Comment:
nit: Considering removing this newline; having a single chunk starting with the tracing message and ending with the assert helps readability in these long tests with multiple sections.